plugins/adr-common/src/index.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * Common types and functionalities for the ADR plugin.
+ * @packageDocumentation
+ */
+import { Entity, getEntitySourceLocation } from '@backstage/catalog-model';
+import { IndexableDocument } from '@backstage/plugin-search-common';
+
+/**
+ * ADR plugin annotation.
+ * @public
+ */
+export const ANNOTATION_ADR_LOCATION_DIR = 'adr/location-dir';
+
+/**
+ * Standard luxon DateTime format string for MADR dates.
+ * @public
+ */
+export const MADR_DATE_FORMAT = 'yyyy-MM-dd';
+
+/**
+ * Utility function to get the value of an entity ADR annotation.
+ * @public
+ */
+export const getAdrLocationDir = (entity: Entity) =>
+  entity.metadata.annotations?.[ANNOTATION_ADR_LOCATION_DIR]?.trim().replace(
+    /^\/+|(?<!\/)\/+$/g,
+    '',
+  );
+
+/**
+ * Utility function to determine if the given entity has ADRs.
+ * @public
+ */
+export const isAdrAvailable = (entity: Entity) =>
+  Boolean(getAdrLocationDir(entity));
+
+/**
+ * Utility function to extract the ADR location URL from an entity based off
+ * its ADR annotation and relative to the entity source location.
+ * @public
+ */
+export const getAdrLocationUrl = (entity: Entity) => {
+  return `${getEntitySourceLocation(entity).target}${getAdrLocationDir(
```

Comment:
Could be best to make sure this is a `url` location perhaps? Seeing as we're passing this to `readTree` anything bug that is likely to explode.

On a separate note, perhaps this should be relative to the entity source location instead? Just so you can allow keeping the ADRs in a separate repo

Reply:
Not quite sure I follow? This is currently relative to the entity source location (did you mean "shouldn't be relative" 😄 )

One thing I considered here is supporting `url:` prefixed annotation values (similar to TechDocs) in the future to allow absolute locations that can point to ADRs in separate repos.

Reply:
I'm thinking the annotation value would be a URL that's resolved relative to the entity source location, so if the annotation is an absolute URL that'd be used directly by itself.

Something like this:

| Source location | Annotation value | Result |
| --------------- | ----------------- | ------ |
| `https://github.com/a/b/tree/main/entities` | `./adrs` | `https://github.com/a/b/tree/main/entities/adrs` |
| `https://github.com/a/b/tree/main/entities/x/y/` | `/adrs` | `https://github.com/a/b/tree/main/adrs` |
| `https://example.com` | `/adrs` | `https://example.com/adrs` |
| `https://example.com` | `https://other-example.com/adrs` | `https://other-example.com/adrs` |

The integrations should help with resolving URLs properly for various providers iirc

Reply:
Makes sense! Updated using `ScmIntegrations`